# poreranger
poreranger is deveopled as the long-read equivalent of cellranger/spaceranger for single-cell and spatial transcriptomics data analysis, as generated by Oxford Nanopore Technologies but also Pacific Biosciences. It has been primarily tested on targeted libraries generated using 10X 5' Chromium and slide-seq 3' platforms. It can be used for immune profiling and genotyping from other library types with minimal modifications.

## Dependencies
### Core programs assumed to be in path:
This program has been tested on Python 3.7.10. Please refer to the provided link for each of the following programs to install them prior to start of your data analysis using this program.

[STAR](https://github.com/alexdobin/STAR) is used to enable core functionality of this program, barcode correction against a set of known barcodes. By certain input parameter changes we use STAR in a Smith-Waterman mode.

[minimap2](https://github.com/lh3/minimap2) is used for initial alignment of raw nanopore reads which subsequently enables recovery of fused/follow-on molecules generated adundantly in newer ONT chemistry LSK112 (aka Q20) by processing all supplemenaty reads. This step also enables orientation, deconcatenation and extraction of subread components (BC-UMI, polyA, etc.).

[SAMtools](http://www.htslib.org/download/) is used for sorting and indexing BAM files

[pigz](https://zlib.net/pigz/) is used for compressing output and intermediate fasta and fastq files.

[MiXCR](https://github.com/milaboratory/mixcr) is used for VDJ alignment and clonotype extraction.

[SeqKit](https://bioinf.shenwei.me/seqkit/) is used for splitting input fastq files in case of very large libraries or libraries prepared with cDNA concatenation. Deconcatenation speed-up is achieved by parallel processing of splitted input files. To enable this step set the optional boolean flag --split.

### External Python packages used:
pysam

edlib

numpy

pandas

matplotlib

seaborn

## Download
```
git clone https://github.com/mehdiborji/poreranger.git

```
## Sample input command
```
python ~/poreranger/pipeline.py  --o outdir --c 4 --i ~/poreranger/data/5mer.gz --e expname  --s
```
